cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataFabricDistributedModule.java

```diff
@@ -57,13 +69,24 @@ public void configure() {
 
     // bind transactions
     bind(TransactionSystemClientService.class).to(DistributedTransactionSystemClientService.class);
-    install(Modules.override(new TransactionModules(txClientId).getDistributedModules()).with(new AbstractModule() {
-      @Override
-      protected void configure() {
-        // Binds the tephra MetricsCollector to the one that emit metrics via MetricsCollectionService
-        bind(MetricsCollector.class).to(TransactionManagerMetricsCollector.class).in(Scopes.SINGLETON);
-      }
-    }));
+
+    // some of these classes need to be non-singleton in order to create a new instance during leader() in
+    // TransactionService
+    bind(SnapshotCodecProvider.class).in(Scopes.SINGLETON);
+    bind(TransactionStateStorage.class).annotatedWith(Names.named("persist")).to(HDFSTransactionStateStorage.class);
+    bind(TransactionStateStorage.class).toProvider(TransactionStateStorageProvider.class);
+    // to catch issues during configure time
+    bind(TransactionManager.class);
+
+    bindConstant().annotatedWith(Names.named(TxConstants.CLIENT_ID)).to(txClientId);
+
+    bind(TransactionSystemClient.class).toProvider(DistributedTxSystemClientProvider.class).in(Scopes.SINGLETON);
+    bind(MetricsCollector.class).to(TransactionManagerMetricsCollector.class).in(Scopes.SINGLETON);
+
+    install(new FactoryModuleBuilder()
+              .implement(TransactionExecutor.class, DefaultTransactionExecutor.class)
+              .build(TransactionExecutorFactory.class));
+
     install(new TransactionExecutorModule());

```

Comment:
Do we still need this?

Reply:
This provides a binding `TransactionExecutorFactory` to build the TransactionExecutor, don't think we should remove this.